File: tree_layout/walker1991.py
# http://www.pennelynn.com/Documents/CUJ/HTML/91HTML/19910056.HTM

import math
from typing import Optional


class Node:
    def __init__(self, value, width=2, parent: Optional["Node"] = None):
        self.val = value
        self.width = width

        # Tree structure
        # This assumes sibling nodes are constructed in the order we want to see them in the tree
        self.parent: Optional["Node"] = parent
        self.level: int = parent.level + 1 if parent is not None else 0
        self.children = []
        self.left_sibling = None
        self.right_sibling = None
        if self.parent:
            if len(self.parent.children) > 0:
                self.left_sibling = self.parent.children[-1]
                self.left_sibling.right_sibling = self
            self.parent.children.append(self)

        # this represents the previous cousin at the same level (depth) of the tree; this could be a sibling, a first
        # cousin or a remote cousin. This is built in the `InitPrevNodeAtLevel` call.
        # this is needed to check and avoid overlaps between subtrees.
        self.left_cousin: Optional["Node"] = None

        # Location parameters; at the end of the tree_position call those will represent top and left of the
        # node (assuming "north" orientation)
        self.x = None
        self.y = None
        # Internal parameters used to calculate the location
        self.prelim = None
        self.modifier = None

    def is_leaf(self):
        return len(self.children) == 0

    def has_left_sibling(self):
        return self.left_sibling is not None

    def has_right_sibling(self):
        return self.right_sibling is not None

    @property
    def first_child(self) -> Optional["Node"]:
        if self.is_leaf():
            return None
        return self.children[0]

    @property
    def last_child(self) -> Optional["Node"]:
        if self.is_leaf():
            return None
        return self.children[-1]

    def __repr__(self):
        return self.val


class Walker1991:
    def __init__(self,
                 sibling_separation=4,
                 maximum_depth=math.inf,
                 root_orientation='NORTH',
                 subtree_separation=4,
                 node_height=2,
                 level_separation=1,
                 min_x=0,
                 min_y=0):
        self.sibling_separation = sibling_separation
        self.maximum_depth = maximum_depth
        self.mean_width = 0
        self.root_orientation = root_orientation
        self.subtree_separation = subtree_separation
        self.node_height = node_height
        self.level_separation = level_separation
        self.min_x = min_x
        self.min_y = min_y
        self.last_node_at_level = {}

    def tree_position(self, apex: Node):
        # Determine the coordinates for each node in a tree.
        # Input: Pointer to the apex node of the tree
        self.last_node_at_level.clear()

        self.tree_first_walk(apex)

        self.tree_second_walk(apex)

        if self.root_orientation in ('NORTH', 'SOUTH'):
            self.fix_min_x(apex)
        else:
            self.fix_min_y(apex)

    def tree_first_walk(self, node: Node, level: int = 0):
        # Clean up old values in a node's flModifier
        node.modifier = 0

        # we build the left cousin pointers on the fly
        node.left_cousin = self.last_node_at_level.get(level)
        self.last_node_at_level[level] = node

        #  In a first post-order walk, every node of the tree is assigned a preliminary x-coordinate (node.prelim).
        #  In addition, internal nodes are given modifiers, which will be used to move their children to the right
        #  (node.modifier).
        if node.is_leaf() or level == self.maximum_depth:
            if node.has_left_sibling():
                # Determine the preliminary x-coordinate based on:
                # - preliminary x-coordinate of left sibling,
                # - the separation between sibling nodes, and
                # - mean width of left sibling & current node.

                #  Set the mean width of these two nodes
                self.tree_mean_node_size(node.left_sibling, node)

                node.prelim = node.left_sibling.prelim + self.sibling_separation + self.mean_width
            else:
                node.prelim = 0
        else:
            #  Position the leftmost of the children
            for child in node.children:
                self.tree_first_walk(child, level + 1)

            #  Calculate the preliminary value between  the children at the far left and right
            midpoint = (node.first_child.prelim + node.last_child.prelim) / 2

            # Set global mean width of these two nodes
            self.tree_mean_node_size(node.left_sibling, node)

            if node.has_left_sibling():
                node.prelim = node.left_sibling.prelim + self.sibling_separation + self.mean_width
                node.modifier = node.prelim - midpoint
                self.tree_apportion(node, level)
            else:
                node.prelim = midpoint

    def tree_mean_node_size(self, left_node, right_node):
        # Write your own code for this procedure if your  rendered nodes will have variable sizes.
        # ----------------------------------------------------------
        # Here I add the width of the contents of the right half of the pLeftNode to the left half of the
        # right node. Since the size of the contents for all nodes is currently the same, this module computes the
        # following trivial computation.
        self.mean_width = 0
        if self.root_orientation in ('NORTH', 'SOUTH'):
            if left_node:
                self.mean_width += left_node.width / 2
            if right_node:
                self.mean_width += right_node.width / 2
        else:
            if left_node:
                self.mean_width += left_node.height / 2
            if right_node:
                self.mean_width += right_node.height / 2

    def tree_apportion(self, node, level):
        #  Clean up the positioning of small sibling subtrees.
        #  Subtrees of a node are formed independently and
        #  placed as close together as possible. By requiring
        #  that the subtrees be rigid at the time they are put
        #  together, we avoid the undesirable effects that can
        #  accrue from positioning nodes rather than subtrees.
        left_most: Node = node.first_child
        compare_depth = 1
        depth_to_stop = self.maximum_depth - level

        # Compute the location of `left_most` and where it should be with respect to `neighbor`.
        while left_most and compare_depth <= depth_to_stop:
            neighbor = left_most.left_cousin if left_most else None

            # no subtree to the left of this subtree from this level onwards
            if neighbor is None:
                break

            right_modsum = left_modsum = 0
            ancestor_leftmost = left_most
            ancestor_neighbor = neighbor
            for _ in range(0, compare_depth):
                ancestor_leftmost = ancestor_leftmost.parent
                ancestor_neighbor = ancestor_neighbor.parent
                right_modsum += ancestor_leftmost.modifier
                left_modsum += ancestor_neighbor.modifier

                # Determine the flDistance to be moved, and apply it to `node` subtree.  Apply appropriate
            # portions to smaller interior subtrees
            self.tree_mean_node_size(left_most, neighbor)

            distance = (neighbor.prelim + left_modsum + self.subtree_separation +
                        self.mean_width - (left_most.prelim + right_modsum))

            if distance > 0:
                # Count the interior sibling subtrees
                n_left_siblings = 0
                sibling: Node = node  # "pTempPtr"
                while sibling is not None and sibling != ancestor_neighbor:
                    sibling = sibling.left_sibling
                    n_left_siblings += 1

                if sibling is None:
                    # Don't need to move anything--it needs to be done by an ancestor because
                    # pAncestorNeighbor and pAncestorLeftmost are not siblings of each other.
                    return

                #  Apply portions to appropriate leftsibling subtrees.
                portion = distance / n_left_siblings
                sibling: Node = node
                while sibling != ancestor_neighbor:
                    sibling.prelim += distance
                    sibling.modifier += distance
                    distance -= portion
                    sibling = sibling.left_sibling

            # Determine the leftmost descendant of `node`  at the next lower level to compare its
            # positioning against that of its `neighbor`
            compare_depth += 1
            if left_most.is_leaf():
                left_most = self.tree_get_leftmost(node, 0, compare_depth)
            else:
                left_most = left_most.first_child

    def tree_get_leftmost(self, node: Node, level: int, search_depth: int):
        #  Determine the leftmost descendant of a node at a given search depth. This is implemented using a post-order
        #  walk of the subtree under `node`, down to the level of `search_depth`. If we've searched to the
        #  proper distance, return the currently leftmost node.
        #  Otherwise, recursively look at the progressively lower levels.
        if level == search_depth:
            return node  # searched far enough.

        # Do a post-order walk of the subtree.
        for child in node.children:
            left_most = self.tree_get_leftmost(child, level + 1, search_depth)
            if left_most:
                return left_most

        # no subtree get to depth search_depth (or there isn't any subtree)
        return None

    def tree_second_walk(self,
                         node: Node,
                         modsum: int = 0):
        # During a second pre-order walk, each node is given a final x-coordinate by summing its preliminary
        # x-coordinate and the modifiers of all the node's ancestors.  The y-coordinate depends on the height of
        # the tree.
        # (The roles of x and y are reversed for RootOrientations of EAST or WEST.)
        if node.level > self.maximum_depth:
            return

        if self.root_orientation == 'NORTH':
            node.x = node.prelim + modsum
            node.y = self.min_y + node.level * (self.node_height + self.level_separation)
        elif self.root_orientation == 'SOUTH':
            node.x = node.prelim + modsum
            node.y = self.min_y + node.level * (self.node_height + self.level_separation)
        elif self.root_orientation == 'EAST':
            node.x = self.min_y + node.level * (self.node_height + self.level_separation)
            node.y = node.prelim + modsum
        else:
            node.x = self.min_y + node.level * (self.node_height + self.level_separation)
            node.y = node.prelim + modsum

        for child in node.children:
            self.tree_second_walk(child, modsum + node.modifier)

    def fix_min_x(self, apex):
        def tree_min(node):
            if node.is_leaf():
                return node.x
            return min(tree_min(child) for child in node.children)

        def shift(node, shift_by):
            node.x += shift_by
            for child in node.children:
                shift(child, shift_by)

        shift(apex, self.min_x - tree_min(apex))

    def fix_min_y(self, apex):
        def tree_min(node):
            if node.is_leaf():
                return node.y
            return min(tree_min(child) for child in node.children)

        def shift(node, shift_by):
            node.y += shift_by
            for child in node.children:
                shift(child, shift_by)

        shift(apex, self.min_x - tree_min(apex))

File: tree_layout/walker1991.c
/***********************************************************
 *  Node-Positioning for General Trees, by John Q. Walker II
 *
 *  Initiated by calling procedure TreePosition().
 **********************************************************/

#include <stdlib.h>

/*----------------------------------------------------------
 * Implementation dependent: Set the values for each of
 * these variables.
 *--------------------------------------------------------*/
#define NODE_WIDTH          20  /* Width of a node?       */
#define NODE_HEIGHT         10  /* Height of a node?      */
#define FRAME_THICKNESS     1   /* Fixed-sized node frame?*/
#define SUBTREE_SEPARATION  5   /* Gap between subtrees?  */
#define SIBLING_SEPARATION  4   /* Gap between siblings?  */
#define LEVEL_SEPARATION    5   /* Gap between levels?    */
#define MAXIMUM_DEPTH       10  /* Biggest tree?          */

/*----------------------------------------------------------
 * Implementation dependent: The structure for one node
 * - The first 4 pointers must be set for each node before
 *   this algorithm is called.
 * - The X and Y coordinates must be set for only the apex
 *   of the tree upon entry; they will be set by this
 *   algorithm for all the other nodes.
 * - The next three elements are used only for the duration
 *   of the algorithm.
 * - Actual contents of the node depend on your application.
 *--------------------------------------------------------*/
typedef int COORD;              /* X,Y coordinate type    */

typedef struct node {
   struct node *parent;        /* ptr: node's parent     */
   struct node *offspring;     /* ptr: leftmost child    */
   struct node *leftsibling;   /* ptr: sibling on left   */
   struct node *rightsibling;  /* ptr: sibling on right  */
   COORD  xCoordinate;         /* node's current x coord */
   COORD  yCoordinate;         /* node's current y coord */

   struct node *prev;          /* ptr: lefthand neighbor */
   float  flPrelim;            /* preliminary x coord    */
   float  flModifier;          /* temporary modifier     */

   char   info[80];            /* pick your contents!    */
} *PNODE;                       /* ptr: a node structure  */

/*----------------------------------------------------------
 * Global variables used by the algorithm
 *--------------------------------------------------------*/
typedef enum { FALSE, TRUE } BOOL;
typedef enum { FRAME, NO_FRAME } FRAME_TYPE;
typedef enum { NORTH, SOUTH, EAST, WEST } ROOT_ORIENTATION;

typedef struct prev_node {      /* one list element        */
   PNODE pPreviousNode;        /* ptr: previous at level  */
   struct prev_node *pNextLevel;   /*  ptr: next element  */
} *PPREVIOUS_NODE;

static FRAME_TYPE FrameType = FRAME;    /*  Show a frame   */
static ROOT_ORIENTATION RootOrientation = NORTH; /*  At top*/
static PPREVIOUS_NODE pLevelZero = (PPREVIOUS_NODE)0;

static COORD xTopAdjustment;    /* How to adjust the apex */
static COORD yTopAdjustment;    /* How to adjust the apex */
static float flMeanWidth;       /* Ave. width of 2 nodes  */

#define FIRST_TIME (0)          /* recursive proc flag    */

/*----------------------------------------------------------
 * Implemented as macros, but could be implemented as
 * procedures depending on your particular node structures
 *--------------------------------------------------------*/

#define FirstChild(node)     ((PNODE)((node)->offspring))
#define LeftSibling(node)    ((PNODE)((node)->leftsibling))
#define RightSibling(node)   ((PNODE)((node)->rightsibling))
#define Parent(node)         ((PNODE)((node)->parent))
#define LeftNeighbor(node)   ((PNODE)((node)->prev))
#define IsLeaf(node)          \
            (((node)&&(!((node)->offspring)))?TRUE:FALSE)
#define HasChild(node)        \
                  (((node)&&((node)->offspring))?TRUE:FALSE)
#define HasLeftSibling(node)   \
             (((node)&&((node)->leftsibling))?TRUE:FALSE)
#define HasRightSibling(node)  \
            (((node)&&((node)->rightsibling))?TRUE:FALSE)

static PNODE GetPrevNodeAtLevel (unsigned nLevelNbr)
{
   /*------------------------------------------------------
    * List Manipulation: Return pointer to previous node at
    * this level
    *----------------------------------------------------*/
   PPREVIOUS_NODE pTempNode;   /* used in the for-loop   */
   unsigned i = 0;             /* level counter          */

   for (pTempNode = pLevelZero; (pTempNode);
       pTempNode = pTempNode->pNextLevel) {
      if (i++ == nLevelNbr)
         /* Reached desired level.  Return its pointer */
         return (pTempNode->pPreviousNode);
   }
   return ((PNODE)0); /* No pointer yet for this level. */
}

static BOOL SetPrevNodeAtLevel (unsigned nLevelNbr,
                           PNODE pThisNode)
{

   /*------------------------------------------------------
    * List Manipulation: Set the list element to the
    * previous node at this level
    *----------------------------------------------------*/

   PPREVIOUS_NODE pTempNode;   /* used in the for-loop   */
   PPREVIOUS_NODE pNewNode;    /* newly-allocated memory */
   unsigned i = 0;             /* level counter          */

   for (pTempNode = pLevelZero; (pTempNode);
       pTempNode = pTempNode->pNextLevel) {
      if (i++ == nLevelNbr) {
         /* Reached desired level. Return its pointer */
         pTempNode->pPreviousNode = pThisNode;
         return (TRUE);
      }
      else if (pTempNode->pNextLevel==(PPREVIOUS_NODE)0) {
         /* Looks like we need a new level: add it.    */
         /* We need to keep going--should be the next. */
         pNewNode = (PPREVIOUS_NODE)
                  malloc(sizeof(struct prev_node));
         if (pNewNode) {
            pNewNode->pPreviousNode = (PNODE)0;
            pNewNode->pNextLevel = (PPREVIOUS_NODE)0;
            pTempNode->pNextLevel = pNewNode;
         }
         else return (FALSE);    /* The malloc failed. */
      }
   }

   /* Should only get here if pLevelZero is 0.           */
   pLevelZero = (PPREVIOUS_NODE)
              malloc(sizeof(struct prev_node));
   if (pLevelZero) {
      pLevelZero->pPreviousNode = pThisNode;
      pLevelZero->pNextLevel = (PPREVIOUS_NODE)0;
      return (TRUE);
   }
   else return (FALSE);        /* The malloc() failed.   */
}

static void InitPrevNodeAtLevel (void)
{
   /*------------------------------------------------------
    * List Manipulation: Initialize the list of the
    * previous node at each level to all zeros.
    *----------------------------------------------------*/

   PPREVIOUS_NODE pTempNode = pLevelZero; /*  the start    */

   for ( ; (pTempNode); pTempNode = pTempNode->pNextLevel)
      pTempNode->pPreviousNode = (PNODE)0;
}

static BOOL CheckExtentsRange(long lxTemp, long lyTemp)
{
   /*-----------------------------------------------------
    * Insert your own code here, to check when the
    * tree drawing is going to be too big. My region is no
    * more that 64000 units square.
    *---------------------------------------------------*/

   if ((labs(lxTemp) > 32000) || (labs(lyTemp) > 32000))
      return (FALSE);
   else
      return (TRUE);
}

static void TreeMeanNodeSize (PNODE pLeftNode,
                         PNODE pRightNode)
{
   /*------------------------------------------------------
    * Write your own code for this procedure if your
    * rendered nodes will have variable sizes.
    *------------------------------------------------------
    * Here I add the width of the contents of the
    * right half of the pLeftNode to the left half of the
    * right node. Since the size of the contents for all
    * nodes is currently the same, this module computes the
    * following trivial computation.
    *----------------------------------------------------*/

   flMeanWidth = (float)0.0;   /* Initialize this global */

   switch (RootOrientation) {
      case NORTH:
      case SOUTH:
         if (pLeftNode) {
            flMeanWidth = flMeanWidth +
                       (float)(NODE_WIDTH/2);
            if (FrameType != NO_FRAME)
               flMeanWidth = flMeanWidth +
                             (float)FRAME_THICKNESS;
         }
         if (pRightNode) {
            flMeanWidth = flMeanWidth +
                       (float)(NODE_WIDTH/2);
            if (FrameType != NO_FRAME)
                flMeanWidth = flMeanWidth +
                              (float)FRAME_THICKNESS;
         }
         break;
      case EAST :
      case WEST :
         if (pLeftNode) {
            flMeanWidth = flMeanWidth +
                       (float)(NODE_HEIGHT/2);
            if (FrameType != NO_FRAME)
                flMeanWidth = flMeanWidth +
                              (float)FRAME_THICKNESS;
         }
         if (pRightNode) {
            flMeanWidth = flMeanWidth +
                       (float)(NODE_HEIGHT/2);
            if (FrameType != NO_FRAME)
                flMeanWidth = flMeanWidth +
                              (float)FRAME_THICKNESS;
         }
         break;
   }
}

static PNODE TreeGetLeftmost(PNODE pThisNode,
                        unsigned nCurrentLevel,
                        unsigned nSearchDepth)
{
   /*------------------------------------------------------
    * Determine the leftmost descendant of a node at a
    * given depth. This is implemented using a post-order
    * walk of the subtree under pThisNode, down to the
    * level of nSearchDepth. If we've searched to the
    * proper distance, return the currently leftmost node.
    * Otherwise, recursively look at the progressively
    * lower levels.
    *----------------------------------------------------*/

   PNODE pLeftmost;    /* leftmost descendant at level   */
   PNODE pRightmost;   /* rightmost offspring in search  */

   if (nCurrentLevel == nSearchDepth)
      pLeftmost = pThisNode; /*  searched far enough.    */
   else if (IsLeaf(pThisNode))
      pLeftmost = 0;  /* This node has no descendants    */
   else {  /* Do a post-order walk of the subtree.        */
      for (pLeftmost = TreeGetLeftmost(pRightmost =
                          FirstChild(pThisNode),
                          nCurrentLevel + 1,
                          nSearchDepth)

         ;
         (pLeftmost==0) && (HasRightSibling(pRightmost))
         ;
         pLeftmost = TreeGetLeftmost(pRightmost =
                          RightSibling(pRightmost),
                          nCurrentLevel + 1,
                          nSearchDepth)
         ) { /* Nothing inside this for-loop. */ }
   }
   return (pLeftmost);
}

static void TreeApportion (PNODE pThisNode,
                      unsigned nCurrentLevel)
{
   /*------------------------------------------------------
    * Clean up the positioning of small sibling subtrees.
    * Subtrees of a node are formed independently and
    * placed as close together as possible. By requiring
    * that the subtrees be rigid at the time they are put
    * together, we avoid the undesirable effects that can
    * accrue from positioning nodes rather than subtrees.
    *----------------------------------------------------*/

   PNODE pLeftmost;            /* leftmost at given level*/
   PNODE pNeighbor;            /* node left of pLeftmost */
   PNODE pAncestorLeftmost;    /* ancestor of pLeftmost  */
   PNODE pAncestorNeighbor;    /* ancestor of pNeighbor  */
   PNODE pTempPtr;             /* loop control pointer   */
   unsigned i;                 /* loop control           */
   unsigned nCompareDepth;     /* depth of comparison    */
                          /* within this proc       */
   unsigned nDepthToStop;      /* depth to halt          */
   unsigned nLeftSiblings;     /* nbr of siblings to the */
           /* left of pThisNode, including pThisNode,  */
           /* til the ancestor of pNeighbor            */
   float flLeftModsum;         /* sum of ancestral mods  */
   float flRightModsum;        /* sum of ancestral mods  */
   float flDistance;           /* difference between     */
        /* where pNeighbor thinks pLeftmost should be   */
        /* and where pLeftmost actually is              */
   float flPortion;            /* proportion of          */
        /* flDistance to be added to each sibling       */

   pLeftmost = FirstChild(pThisNode);
   pNeighbor = LeftNeighbor(pLeftmost);

   nCompareDepth = 1;
   nDepthToStop = MAXIMUM_DEPTH - nCurrentLevel;

   while ((pLeftmost) && (pNeighbor) &&
         (nCompareDepth <= nDepthToStop)) {

      /* Compute the location of pLeftmost and where it */
      /* should be with respect to pNeighbor.           */
      flRightModsum = flLeftModsum = (float)0.0;
      pAncestorLeftmost = pLeftmost;
      pAncestorNeighbor = pNeighbor;
      for (i = 0; (i < nCompareDepth); i++) {
         pAncestorLeftmost = Parent(pAncestorLeftmost);
         pAncestorNeighbor = Parent(pAncestorNeighbor);
         flRightModsum = flRightModsum +
                      pAncestorLeftmost->flModifier;
         flLeftModsum = flLeftModsum +
                      pAncestorNeighbor->flModifier;

      }

      /* Determine the flDistance to be moved, and apply*/
      /* it to "pThisNode's" subtree.  Apply appropriate*/
      /* portions to smaller interior subtrees          */

      /* Set the global mean width of these two nodes   */
      TreeMeanNodeSize(pLeftmost, pNeighbor);

      flDistance = (pNeighbor->flPrelim +
                  flLeftModsum +
                  (float)SUBTREE_SEPARATION +
                  (float)flMeanWidth) -
                 (pLeftmost->flPrelim + flRightModsum);

      if (flDistance > (float)0.0) {
         /* Count the interior sibling subtrees        */
         nLeftSiblings = 0;
         for (pTempPtr = pThisNode;
              (pTempPtr) &&
              (pTempPtr != pAncestorNeighbor);
             pTempPtr = Leftsibling(pTempPtr)) {
            nLeftSiblings++;
         }

         if (pTempPtr) {
            /* Apply portions to appropriate          */
            /* leftsibling subtrees.                  */
            flPortion = flDistance/(float)nLeftSiblings;
            for (pTempPtr = pThisNode;
                (pTempPtr != pAncestorNeighbor);
                pTempPtr = LeftSibling(pTempPtr)) {
               pTempPtr->flPrelim =
                    pTempPtr->flPrelim + flDistance;
               pTempPtr->flModifier =
                    pTempPtr->flModifier + flDistance;
               flDistance = flDistance - flPortion;
             }
         }
         else {
            /* Don't need to move anything--it needs  */
            /* to be done by an ancestor because      */
            /* pAncestorNeighbor and                  */
            /* pAncestorLeftmost are not siblings of  */
            /* each other.                            */
            return;
         }
      }  /* end of the while                           */

      /* Determine the leftmost descendant of pThisNode */
      /* at the next lower level to compare its         */
      /* positioning against that of its pNeighbor.     */

      nCompareDepth++;
      if (IsLeaf(pLeftmost))
         pLeftmost = TreeGetLeftmost(pThisNode, 0,
                                 nCompareDepth);
      else
         pLeftmost = FirstChild(pLeftmost);
      pNeighbor = LeftNeighbor(pLeftmost);
   }
 }

 static BOOL TreeFirstWalk(PNODE pThisNode,
                       unsigned nCurrentLevel)
 {
   /*------------------------------------------------------
    * In a first post-order walk, every node of the tree is
    * assigned a preliminary x-coordinate (held in field
    * node->flPrelim). In addition, internal nodes are
    * given modifiers, which will be used to move their
    * children to the right (held in field
    * node->flModifier).
    * Returns: TRUE if no errors, otherwise returns FALSE.
    *----------------------------------------------------*/

   PNODE pLeftmost;            /* left- & rightmost      */
   PNODE pRightmost;           /* children of a node.    */
   float flMidpoint;           /* midpoint between left- */
                          /* & rightmost children   */

   /* Set up the pointer to previous node at this level  */
   pThisNode->prev = GetPrevNodeAtLevel(nCurrentLevel);

   /* Now we're it--the previous node at this level      */
   if (!(SetPrevNodeAtLevel(nCurrentLevel, pThisNode)))
      return (FALSE);        /* Can't allocate element  */

   /* Clean up old values in a node's flModifier         */
   pThisNode->flModifier = (float)0.0;

   if ((IsLeaf(pThisNode)) ||
      (nCurrentLevel == MAXIMUM_DEPTH)) {
      if (HasLeftSibling(pThisNode)) {

      /*--------------------------------------------
       * Determine the preliminary x-coordinate
       *   based on:
       * - preliminary x-coordinate of left sibling,
       * - the separation between sibling nodes, and
       * - mean width of left sibling & current node.
       *--------------------------------------------*/
      /* Set the mean width of these two nodes      */
      TreeMeanNodeSize(LeftSibling(pThisNode),
                    pThisNode);

      pThisNode->flPrelim =
                (pThisNode->Leftsibling->flPrelim) +
                (float)SIBLING_SEPARATION +
                flMeanWidth;
   }
   else    /*  no sibling on the left to worry about  */
      pThisNode->flPrelim = (float)0.0;
}
else {
   /* Position the leftmost of the children          */
   if (TreeFirstWalk(pLeftmost = pRightmost =
                  FirstChild(pThisNode),
                  nCurrentLevel + 1)) {
      /* Position each of its siblings to its right */
      while (HasRightSibling(pRightmost)) {
         if (TreeFirstWalk(pRightmost =
                        RightSibling(pRightmost),
                        nCurrentLevel + 1)) {
         }
         else return (FALSE); /* malloc() failed   */
      }

      /* Calculate the preliminary value between   */
      /* the children at the far left and right    */
      flMidpoint = (pLeftmost->flPrelim +
                  pRightmost->flPrelim)/(2.0);

      /* Set global mean width of these two nodes  */
      TreeMeanNodeSize(LeftSibling(pThisNode),
                    pThisNode);

         if (HasLeftSibling(pThisNode)) {
            pThisNode->flPrelim =
                   (pThisNode->leftsibling->flPreLim) +
                           (float)SIBLING_SEPARATION +
                           flMeanWidth;
            pThisNode->flModifier =
                pThisNode->flPrelim - flMidpoint;
            TreeApportion(pThisNode, nCurrentLevel);
         }
         else pThisNode->flPrelim = flMidpoint;
      }
      else return (FALSE); /* Couldn't get an element  */
   }
   return (TRUE);
}

static BOOL TreeSecondWalk(PNODE pThisNode,
                      unsigned nCurrentLevel)
{
   /*------------------------------------------------------
    * During a second pre-order walk, each node is given a
    * final x-coordinate by summing its preliminary
    * x-coordinate and the modifiers of all the node's
    * ancestors.  The y-coordinate depends on the height of
    * the tree.  (The roles of x and y are reversed for
    * RootOrientations of EAST or WEST.)
    * Returns: TRUE if no errors, otherwise returns FALSE.
    *----------------------------------------- ----------*/

   BOOL bResult = TRUE;        /* assume innocent        */
   long lxTemp, lyTemp;        /* hold calculations here */
   float flNewModsum;          /* local modifier value   */
   static float flModsum = (float)0.0;

   if (nCurrentLevel <= MAXIMUM_DEPTH) {
      flNewModsum = flModsum;  /* Save the current value  */
      switch (RootOrientation) {
         case NORTH:
            lxTemp = (long)xTopAdjustment + (long)(pThisNode->flPrelim + flModsum);
            lyTemp = (long)yTopAdjustment + (long)(nCurrentLevel * LEVEL_SEPARATION);
            break;
         case SOUTH:
            lxTemp = (long)xTopAdjustment + (long)(pThisNode->flPrelim + flModsum);
            lyTemp = (long)yTopAdjustment - (long)(nCurrentLevel * LEVEL_SEPARATION);
            break;
         case EAST:
            lxTemp = (long)xTopAdjustment + (long)(nCurrentLevel * LEVEL_SEPARATION);
            lyTemp = (long)yTopAdjustment - (long)(pThisNode->flPrelim + flModsum);
            break;
         case WEST:
            lxTemp = (long)xTopAdjustment - (long)(nCurrentLevel * LEVEL_SEPARATION);
            lyTemp = (long)yTopAdjustment - (long)(pThisNode->flPrelim + flModsum);
            break;
      }

      if (CheckExtentsRange(lxTemp, lyTemp)) {
         /* The values are within the allowable range */

         pThisNode->xCoordinate = (COORD)lxTemp;
         pThisNode->yCoordinate = (COORD)lyTemp;

         if (HasChild(pThisNode)) {
            /* Apply the flModifier value for this    */
            /* node to all its offspring.             */
            flModsum = flNewModsum =
                   flNewModsum + pThisNode->flModifier;
            bResult = TreeSecondWalk(
               FirstChild(pThisNode),nCurrentLevel + 1);
            flNewModsum = flNewModsum -
                        pThisNode->flModifier;
         }

         if ((HasRightSibling(pThisNode)) && (bResult)) {
            flModsum = flNewModsum;
            bResult = TreeSecondWalk(
                RightSibling(pThisNode), nCurrentLevel);
         }
      }
      else bResult = FALSE;   /* outside of extents   */
   }
   return (bResult);
}

static BOOL TreePosition(PNODE pApexNode)
{
   /*------------------------------------------------------
    * Determine the coordinates for each node in a tree.
    * Input: Pointer to the apex node of the tree
    * Assumption: The x & y coordinates of the apex node
    * are already correct, since the tree underneath it
    * will be positioned with respect to those coordinates.
    * Returns: TRUE if no errors, otherwise returns FALSE.
    *----------------------------------------------------*/

   if (pApexNode) {
      /* Initialize list of previous node at each level */
      InitPrevNodeAtLevel();

      /* Generate the properly-placed tree nodes.      */
      /* TreeFirstWalk: a post-order walk              */
      /* TreeSecondWalk: a pre-order walk              */
      if (TreeFirstWalk (pApexNode, FIRST_TIME)) {
         /* Determine how to adjust the nodes with     */
         /* respect to the location of the apex of the */
         /* tree being positioned.                     */
         switch (RootOrientation) {
            case NORTH:
            case SOUTH:
               /* Create the adjustment from x-coord  */
               xTopAdjustment = pApexNode->xCoordinate -
                        (COORD)(pApexNode->flPrelim);
               yTopAdjustment = pApexNode->yCoordinate;
               break;
            case EAST :
            case WEST :
               /* Create the adjustment from y-coord  */
               xTopAdjustment = pApexNode->xCoordinate;
               yTopAdjustment = pApexNode->yCoordinate +
                        (COORD)(pApexNode->flPrelim);
               break;
         }
         return (TreeSecondWalk(pApexNode, FIRST_TIME));
      }
      else return (FALSE); /*  Couldn't get an element   */
   }
   else return (TRUE); /*  Easy: null pointer was passed  */
}